### ISSUE TITLE
What are costs to agencies enabling HTTPS for their websites?  

### ISSUE THREAD

**Issue by gbinal** [GSA/https#4]
For discussion...

**Comment by gbinal** [comment]
> The annual cost of procuring a certificate for a single domain can range from 1 digit ([seriously](https://www.namecheap.com/security/ssl-certificates/comodo/positivessl.aspx)) to 4 digits. 

> When 18F buys its own certs, we spend $16 per year per single domain (e.g. myra.treasury.gov), or $150 per year for a cert that can cover a range of subdomains (e.g. *.18f.gov). We find this a minimal cost.

> Some organizations favor a more expensive kind of certificate called "extended validation", which causes browsers to show the company's full name in the location bar. Twitter does this. This can cost in the 4-digit range, because it involves human effort to confirm business paperwork. 

> However, not everyone thinks these expensive certs are valuable -- Google and Facebook do not do it, and 18F does not do it. Nonetheless, some organizations feel it is valuable, and would view this as a mandatory cost. I encourage OMB not to view it as a mandatory cost, but to anticipate this reaction.

> In any case, for some agencies with serious domain and subdomain sprawl, they may find certificate procurement adds up, but on a site by site basis, it can and should be a miniscule fraction of the cost of hosting a website.


> As for developer time: the operational cost is more about administration and deployment than developer time, though for larger sites some developer time may be required to update links to old resources. It's very hard to quantify in a general way, because it depends on the site's scope.

> The deployment work can also vary. For example, I've been working with Paul Donohue over on the OMB MAX team to turn on HTTPS for existing MAX-managed domains, and it's been easy and free (MAX can flip a switch). And at 18F (where we make new sites, not retrofit old ones), the deployment work is minimal and a standard part of our operations.

> Agencies who self-manage their systems have varying levels of internal chaos that may increase the time and cost. For example, AIDS.gov has been tackling this with vigor, in the wake of some bad press around not using HTTPS, and it's involved dedicated methodical work looking for the gaps and filling them.

> But to Ben's point: many .gov websites use CDN services (like Akamai), and many CDNs charge extra for HTTPS. I've spoken with other .gov's and they've cited an increased CDN bill as part of their reticence. GSA does have some vehicles for this, though I'm not sure of the details. It's important to note that using a CDN doesn't *require* increased cost -- for example, [CloudFlare](https://www.cloudflare.com/plans) is a popular, world class CDN that offers HTTPS at no extra cost. However, I'm fairly certain no vehicle or precedent exists yet for CloudFlare inside .gov.

> Operational and CDN costs are the kinds of costs that will go down as a result of increased investment and pressure for .gov's to turn on HTTPS, so I see a mandate -- with accompanying OMB guidance and an expression of willingness to work through issues with agencies -- as a big help to changing the landscape here.

**Comment by konklone** [comment]
(For context - that is an email I wrote.)

**Comment by gbinal** [comment]
Another earlier comment:  

> For mandating this for new web sites, the costs are trivial and mostly just the cost of the certificate, which doesn’t need to be expensive (doesn’t the US government have its own Certificate Authority anyway?).
 
> For mandating this for existing web sites that don’t currently have an HTTPS presence, the migration will need to be deliberate and more expensive.  You’d probably want to set up the HTTPS site in parallel, test that everything works, and then convert the HTTP side to a service that just issues redirects to the HTTPS side.  This lets non-HSTS browsers continue to work without breaking everyone’s bookmarks or links on other sites.  Long-term you could fix the links and turn off the HTTP service.
 
> Existing web sites that have non-equivalent HTTP and HTTPS sites would have the added expense of having to confirm the path namespaces on either side don’t conflict, and where they do, sort that out first.
 
> If the existing HTTP site is hosted on infrastructure that relies on IP-less virtual hosting (that is, multiple web sites are served from the same IP address), that will probably have to change (older browsers would need to access different SSL sites on different IP addresses), and so there may be IP address allocation and configuration costs.  Alternatively, you could just decide not to support browsers that don’t support Server Name Indication (SNI).

**Comment by dnesting** [comment]
(and that was mine)

The potentially largest expense I think is not in the configuration or certificate (as others have said), but probably the content changes on the site to ensure no mixed content and to ensure links work.  For web applications built by 3rd parties, this may necessitate some rework of the application.  This seems like it should be easy to do, but I don't know how that translates to costs for the government.

**Comment by konklone** [comment]
Closing in favor of addressing cost in a "certificate issuance" document, which would also cover Domain Validation vs Organizational Validation vs Extended Validation.
